### PR TITLE
refactor: guidelines 3.0 input style 

### PR DESCRIPTION
**Pull request by annsch**
## Additionally
- [x] `docs/migrationGuide.adoc`
- [x] Test-URLs: `.pa11yci` and `tests/backstop.json`
- [x] clean scss

Resolves https://github.com/db-ui/core/issues/223

### CODE REVIEW

File: .htmlvalidate.json [hunk 1]
```diff
@@ -5,6 +5,7 @@
 		"no-trailing-whitespace": ["off"],
 		"doctype-style": ["off"],
 		"element-name": ["error", { "whitelist": ["dialog"] }],
-		"missing-doctype": ["off"]
+		"missing-doctype": ["off"],
+		"no-dup-id": ["off"]
```
Comment (mfranzke) [inline .htmlvalidate.json:9]: let's fix to have duplicated IDs within our examples instead.

Comment (mfranzke) [reply .htmlvalidate.json:9]: I've provided some solution to ensure unique IDs (at least for the iterations for the moment).

File: source/_patterns/01-elements/input/input.demonstration.scss [hunk 2]
```diff
@@ -1,9 +1,21 @@
...
 	}
 }
 
 .elm-input {
 	max-width: 30rem;
 }
+
+.pl-c-pattern {
+	@extend %db-ui-regular;
+}
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.demonstration.scss:21]: what about moving this to `source/css/pattern-scaffolding-project-specific.scss` instead as it's a quite global selector?

File: source/_patterns/01-elements/input/input.hbs [hunk 3]
```diff
@@ -1,25 +1,33 @@
...
+		{{{ validation }}}
+		{{#if datalist }} list="datalist{{ id }}"{{/if }}
+		aria-labelledby="{{ id }}-label">
+	<label for="{{ id }}"
+		aria-hidden="true"
+		id="{{ id }}-label">{{ label }}</label>
+
+	{{#if description }}<p{{#if describedbyid }} id="{{ describedbyid }}"{{/if }} class="description">{{{ description }}}</p>{{/if }}
+
+	{{!-- if span elm-icon o.ä. mit data-icon --}}
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.hbs:25]: is this comment related to code line 32 instead ?

File: source/_patterns/01-elements/input/_input.variables.scss [hunk 4]
```diff
@@ -1,5 +1,15 @@
...
+
+	transition: opacity 0.08s, transform 0.15s;
+	opacity: var(--db-input-focus-opacity);
+
+	transform: translate(
+			$db-spacing-fixed-xs,
+			calc($db-sizing-md * $transformScale)
+		)
+		scale(var(--db-input-label-scale));
+}
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/_input.variables.scss:15]: This seems to have been replaced by `label-focus-animation` within `source/_patterns/01-elements/input/input.scss`, hasn't it ?

File: source/_patterns/01-elements/input/_input.variables.scss [hunk 5]
```diff
@@ -1,5 +1,15 @@
-$icons-path: "../../../icons/" !default;
-@use "@db-ui/base/build/scss/variables" as *;
-@use "@db-ui/base/build/scss/helpers/functions" as *;
-
 $input---color: $db-colors-secondary-enabled !default;
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/_input.variables.scss:1]: this color doesn't seem to be in use anymore, so probably this file could get removed in total.

File: source/_patterns/01-elements/input/input.hbs [hunk 6]
```diff
@@ -1,25 +1,33 @@
...
+	{{#if datalist }}<datalist id="datalist-{{ id }}">
+		{{#each datalistItems }}
+			<option{{#if value }} value="{{ value }}"{{/if }}>{{ text }}</option>
+		{{/each }}
+	</datalist>
+	{{/if }}
+	<span class="elm-state-icon"></span>
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.hbs:32]: Couldn't we use the `icon()` SCSS mixin related to the `data-variant` value on the surrounding `div.elm-input` instead of adding an additional HTML tag ?

File: source/_patterns/01-elements/input/input.scss [hunk 7]
```diff
@@ -1,50 +1,145 @@
...
+		scale(var(--db-input-label-scale));
+}
 
 .elm-input {
-	@extend %form-element;
+	position: relative;
 
-	color: $input---color;
+	--elm-input-padding-left: #{$db-spacing-fixed-sm};
+	--elm-input-padding-right: #{$db-spacing-fixed-sm};
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.scss:26]: let's start to use the new semantic names for padding and margin instead of `left` and `right`:
https://developer.mozilla.org/en-US/docs/Web/CSS/padding-inline-start

File: source/_patterns/01-elements/input/input.scss [hunk 8]
```diff
@@ -1,50 +1,145 @@
...
+	&::after {
+		right: 0;
+	}
+}
 
-	padding: $db-spacing-fixed-regular-xs $db-spacing-fixed-regular-m;
+.elm-input input {
+	@extend %db-bg-neutral-transparent-semi;
 
+	border: none;
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.scss:82]: If we don't need a border, we should use `border-color: transparent` instead for Windows High Contrast mode reasons: https://piccalil.li/quick-tip/use-transparent-borders-and-outlines-to-assist-with-high-contrast-mode/

File: source/_patterns/01-elements/input/input.scss [hunk 9]
```diff
@@ -1,50 +1,145 @@
...
 	fieldset:disabled & {
 		&,
-		& + .elm-label {
+		& + label {
 			// Decided against cursor: not-allowed, compare to e.g. https://phabricator.wikimedia.org/T121960
-			opacity: 0.4;
+			opacity: 0.25;
+		}
+
+		& + .description {
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.scss:141]: In between the input and the description there's most likely the label so we would need the tilde instead of plus here, wouldn't we ? And is this an actual decision to visually hide the text, but keep it's space ?

File: source/_patterns/01-elements/input/input.scss [hunk 10]
```diff
@@ -1,50 +1,145 @@
...
+		align-items: center;
+		font-size: var(--db-type-body-font-size-md);
+		color: $db-colors-secondary-enabled;
+
+		transform-origin: left;
+		transform: translate(
+			var(--elm-input-padding-left),
+			calc(-1 * var(--db-sizing-md))
+		);
+		transition: transform 0.2s $db-transition-functional-timing;
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.scss:102]: We've had the transition limited to specific input types by the declaration at code line https://github.com/db-ui/core/pull/198/files#diff-f6ccff0b7919f06f84b6c7b7f73f5a367782f1b2c95321130b6168c478485a2eL80

File: source/_patterns/01-elements/input/input.demonstration.scss (hunk 2 above)
Comment (annsch) [reply source/_patterns/01-elements/input/input.demonstration.scss]: good point, changed that.

File: source/_patterns/01-elements/input/input.hbs (hunk 3 above)
Comment (annsch) [reply source/_patterns/01-elements/input/input.hbs]: changed in https://github.com/db-ui/core/pull/198/commits/9291c0e751599052a5c87dc858f92b0e719c1eaa

File: source/_patterns/01-elements/input/_input.variables.scss (hunk 4 above)
Comment (annsch) [reply source/_patterns/01-elements/input/_input.variables.scss]: updated in https://github.com/db-ui/core/pull/198/commits/9291c0e751599052a5c87dc858f92b0e719c1eaa

File: source/_patterns/01-elements/input/_input.variables.scss (hunk 5 above)
Comment (annsch) [reply source/_patterns/01-elements/input/_input.variables.scss]: updated in https://github.com/db-ui/core/pull/198/commits/9291c0e751599052a5c87dc858f92b0e719c1eaa

File: source/_patterns/01-elements/input/input.scss (hunk 7 above)
Comment (annsch) [reply source/_patterns/01-elements/input/input.scss]: chaned in https://github.com/db-ui/core/pull/198/commits/9291c0e751599052a5c87dc858f92b0e719c1eaa

File: source/_patterns/01-elements/input/input.scss (hunk 8 above)
Comment (annsch) [reply source/_patterns/01-elements/input/input.scss]: changed in https://github.com/db-ui/core/pull/198/commits/9291c0e751599052a5c87dc858f92b0e719c1eaa

File: source/_patterns/01-elements/input/input.scss (hunk 9 above)
Comment (annsch) [reply source/_patterns/01-elements/input/input.scss]: changed in https://github.com/db-ui/core/pull/198/commits/9291c0e751599052a5c87dc858f92b0e719c1eaa

File: source/_patterns/01-elements/input/input.scss (hunk 10 above)
Comment (annsch) [reply source/_patterns/01-elements/input/input.scss:116]: as discussed, we will not limit the animations as done before according to guidelines 3.0

File: source/_patterns/01-elements/input/input.scss [hunk 11]
```diff
@@ -55,114 +150,197 @@
...
+		~ .description {
+			padding-left: $paddingLeft;
+		}
+
+		&:focus,
+		&:not(:placeholder-shown) {
+			&:not(:disabled) {
+				+ label {
+					padding-left: 0;
+					@include label-focus-animation($paddingLeft);
```
[the remark this answers is not in the document]
Comment (annsch) [reply source/_patterns/01-elements/input/input.scss]: changed that in https://github.com/db-ui/core/pull/198/commits/9291c0e751599052a5c87dc858f92b0e719c1eaa

File: source/_patterns/01-elements/input/input.scss [hunk 12]
```diff
@@ -55,114 +150,197 @@
...
+			pointer-events: all;
+		}
 
-			&:invalid {
-				&:not(:placeholder-shown) {
-					--formElement---borderColor: #{$db-colors-primary-enabled};
-				}
+		&.hide-native-cancel {
+			&::-webkit-search-cancel-button {
+				display: none;
```
[the remark this answers is not in the document]
Comment (annsch) [reply source/_patterns/01-elements/input/input.scss]: removed that in https://github.com/db-ui/core/pull/198/commits/9291c0e751599052a5c87dc858f92b0e719c1eaa

File: source/_patterns/01-elements/input/input.hbs (hunk 6 above)
Comment (annsch) [reply source/_patterns/01-elements/input/input.hbs:33]: As discussed we moved this discussion to https://github.com/db-ui/core/issues/223

File: source/_patterns/01-elements/input/input.hbs [hunk 13]
```diff
@@ -1,25 +1,34 @@
...
+
+	{{#if datalist }}<datalist id="datalist-{{ id }}">
+		{{#each datalistItems }}
+			<option{{#if value }} value="{{ value }}"{{/if }}>{{ text }}</option>
+		{{/each }}
+	</datalist>
+	{{/if }}
+
+	{{!-- icon for data-variants like error or success --}}
+	<span class="elm-state-icon"></span>
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.hbs:33]: Let's try to avoid this tag and use the existing wrapping one within another evaluation.

File: source/_patterns/01-elements/input/input.scss [hunk 14]
```diff
@@ -1,50 +1,159 @@
...
+		--elm-input-padding-inline-end: #{calc(
+				2 * #{$db-spacing-fixed-sm} + 0.425rem + 0.75rem + #{$db-spacing-fixed-xs}
+			)};
 
+		input {
+			padding-inline-end: var(--elm-input-padding-inline-end);
+		}
+
+		.elm-state-icon::after {
+			margin-right: calc(0.75rem + $db-spacing-fixed-xs);
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.scss:60]: You could use `margin-inline-end` as an equivalent to `padding-inline-end` here as well.

File: tests/backstop.json [hunk 15]
```diff
@@ -88,23 +88,28 @@
...
 		{
-			"label": "Elements – Input",
-			"url": "http://127.0.0.1:8080/patterns/elements-input/index.html",
+			"label": "Elements – Input Sizes",
+			"url": "http://127.0.0.1:8080/patterns/elements-input-input-sizes/elements-input-input-sizes.rendered.html",
+			"referenceUrl": "https://db-ui.github.io/core/patterns/elements-input/index.html"
+		},
+		{
+			"label": "Elements – Input Variations",
+			"url": "http://127.0.0.1:8080/patterns/elements-input-input-variations/elements-input-input-variations.rendered.html",
```
Comment (mfranzke) [inline tests/backstop.json:97]: let's not include the different pseudo-patterns, but only test the general page from now on to reduce the test cases, meaning to revert adding the other pages to this file, BUT we would still need to change selectors for the interactions as you did for "Elements – Input - focused" (and we need it for "Elements – Input / hovered" as well).

File: source/_patterns/01-elements/textarea/textarea.scss [hunk 16]
```diff
@@ -3,14 +3,13 @@
 @use "@db-ui/base/build/scss/variables" as *;
 @use "@db-ui/base/build/scss/helpers/functions" as *;
 @import "textarea.variables";
-@import "../form-elements";
 
 .elm-textarea {
-	@extend %form-element;
+	@extend %form-element !optional;
```
Comment (mfranzke) [inline source/_patterns/01-elements/textarea/textarea.scss:8]: do we really need the optional here? I like the idea that the build breaks in case that the relevant styles aren't present, because elsewhere we would get a "valid" build but an unstyled element, especially as we don't have some kind of a fallback, wouldn't we ?

File: source/_patterns/01-elements/input/input.scss (hunk 14 above)
Comment (annsch) [reply source/_patterns/01-elements/input/input.scss]: @mfranzke done

File: tests/backstop.json (hunk 15 above)
Comment (annsch) [reply tests/backstop.json]: done

File: source/_patterns/01-elements/textarea/textarea.scss (hunk 16 above)
Comment (annsch) [reply source/_patterns/01-elements/textarea/textarea.scss]: done and deleted all references to %form-element. Left comment in _textarea.scss_ for later

File: source/_patterns/01-elements/input/input.scss [hunk 17]
```diff
@@ -1,50 +1,159 @@
...
+	&:has(> input:disabled) {
+		&::before,
+		&::after,
+		.elm-state-icon::after {
+			opacity: 0.5;
+		}
+	}
+}
+
+.elm-input input {
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.scss:93]: You could as well wrap this into the previous section / selector.
```suggestion
    input {
```

File: source/_patterns/01-elements/input/input.scss [hunk 18]
```diff
@@ -1,50 +1,159 @@
...
+
+		transform-origin: left;
+		transform: translate(
+			var(--elm-input-padding-inline-start),
+			calc(-1 * var(--db-sizing-md))
+		);
+		transition: transform 0.2s $db-transition-functional-timing;
+
+		height: $db-sizing-md;
+		pointer-events: none;
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.scss:119]: this declaration prevents the clicking / pointing on the `label` and `focus`ing the input element afterwards (even after the label floated away). Do we really need this declaration ?

Comment (mfranzke) [inline source/_patterns/01-elements/input/input.scss:119]: this declaration prevents the clicking / pointing on the `label` and `focus`ing the input element afterwards (even after the label floated away). Do we really need this declaration ?
```suggestion

```

File: source/_patterns/01-elements/input/input.scss [hunk 19]
```diff
@@ -1,50 +1,160 @@
...
-		& + .elm-label {
+		& + label {
 			// Decided against cursor: not-allowed, compare to e.g. https://phabricator.wikimedia.org/T121960
-			opacity: 0.4;
+			opacity: 0.25;
+		}
+
+		& ~ .description {
+			opacity: 0;
 		}
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.scss:158]: I'm not sure, whether I've asked this before, but why would we want to visually hide the description for deactivated fields?

File: source/_patterns/01-elements/input/input.scss [hunk 20]
```diff
@@ -55,114 +165,193 @@
...
+			right: $db-spacing-fixed-sm;
+			top: 0;
+			height: $db-sizing-md;
+			width: 1em;
+			background: url(#{$icons-path}functional/images/navigation/db_ic_cancel_24.svg)
+				no-repeat 50% 50%;
+			background-size: contain;
+			opacity: 0;
+			pointer-events: none;
+			cursor: pointer;
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.scss:245]: I like

File: source/_patterns/01-elements/input/input.scss [hunk 21]
```diff
@@ -55,114 +165,193 @@
...
+			position: absolute;
+			right: $db-spacing-fixed-sm;
+			top: 0;
+			height: $db-sizing-md;
+			width: 1em;
+			background: url(#{$icons-path}functional/images/navigation/db_ic_cancel_24.svg)
+				no-repeat 50% 50%;
+			background-size: contain;
+			opacity: 0;
+			pointer-events: none;
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.scss:244]: why do we need to deactivate the `pointer-events` here? From my understanding this is the browser built-in element and that for I would expect that we actually need its functionality.
```suggestion

```

File: source/_patterns/01-elements/input/input.scss [hunk 22]
```diff
@@ -55,114 +165,193 @@
...
+			cursor: pointer;
 		}
 
-		// ### Style variations
-		&[data-variant="semitransparent"] {
-			@extend %form-element-semitransparent;
+		&:focus::-webkit-search-cancel-button {
+			opacity: 1;
+			pointer-events: all;
```
Comment (mfranzke) [inline source/_patterns/01-elements/input/input.scss:250]: why do we need to deactivate the `pointer-events` and toggle `opacity` here? From my understanding this is the browser built-in element and that for I would expect that we actually need its functionality.
```suggestion

```

File: source/_patterns/01-elements/input/input.scss (hunk 18 above)
Comment (annsch) [reply source/_patterns/01-elements/input/input.scss:119]: @mfranzke by removing this we allow focus on the label and due to the fact, that our label sits always above the input element we'll loose the cursor: text property on hovering.
We do not have the "label next to input element" anymore, so I suggest to leave it with pointer-events: none.

File: source/_patterns/01-elements/input/input.scss (hunk 22 above)
Comment (annsch) [reply source/_patterns/01-elements/input/input.scss:250]: Fine for me to use native behaviour. This leads to showing this Pseudo-Button on hover as well.

File: source/_patterns/01-elements/input/input.scss (hunk 19 above)
Comment (annsch) [reply source/_patterns/01-elements/input/input.scss:158]: removed that. Could be some inherited and not questionated property ;)

File: source/_patterns/01-elements/input/input.scss (hunk 17 above)
Comment (annsch) [reply source/_patterns/01-elements/input/input.scss:93]: done

File: source/_patterns/01-elements/input/input.scss (hunk 18 above)
Comment (annsch) [reply source/_patterns/01-elements/input/input.scss:107]: because of your holidays I suggest to discuss this later; I updated the follow-up issue: https://github.com/db-ui/core/issues/223

**Review by mfranzke — APPROVED** [review]
